### PR TITLE
feat: form/cursor api

### PR DESCRIPTION
**Pull request by armed**
related to #10 

Aimed to provide easy to use API which allows us to replicate vim-sexp functionality without explicitly adding it to plugin.

current state: element wraps and form wraps

Keybinding examples (as in vim-sexp):
```lua
      ["<localleader>w"] = {
        function()
          -- place cursor API, which accepts form and an options (placement type, mode)
          paredit.cursor.place_cursor(
            -- gets element under cursor, finds nearest form and
            -- wraps it with provided prefix and suffix args
            paredit.form.wrap_element_under_cursor("( ", ")"),
            { placement = "inner_start", mode = "insert" }
          )
        end,
        "Wrap element insert head",
      },

      ["<localleader>W"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.form.wrap_element_under_cursor("(", ")"),
            { placement = "inner_end", mode = "insert" }
          )
        end,
        "Wrap element insert tail",
      },

      ["<localleader>i"] = {
        function()
          paredit.cursor.place_cursor(
            -- finds nearest grandparent form
            paredit.form.wrap_enclosing_form_under_cursor("( ", ")"),
            { placement = "inner_start", mode = "insert" }
          )
        end,
        "Wrap form insert head",
      },

      ["<localleader>I"] = {
        function()
          paredit.cursor.place_cursor(
            paredit.form.wrap_enclosing_form_under_cursor("(", ")"),
            { placement = "inner_end", mode = "insert" }
          )
        end,
        "Wrap form insert tail",
      },
```

### CODE REVIEW

File: lua/nvim-paredit/lang/clojure.lua [hunk 1]
```diff
@@ -36,6 +36,16 @@ function M.get_node_root(node)
...
+function M.element_lit(node)
+  local lit = node
+  if lit then
+    if not lit:type():match("lit$") then
+      lit = M.element_lit(lit:parent())
+    end
+  end
+  return lit
+end
+
```
Comment (julienvincent) [inline lua/nvim-paredit/lang/clojure.lua:48]: From what I can see this is trying to solve the same problem as the existing `get_not_root` fn. Unless I am mistaken, we don't need to introduce this new language API and can just use the existing fn.

To clarify this fn will find the root most node relative to another node.

`get_node_root(keyword_node, relative_parent)` -> `namespaced_keyword`.

File: lua/nvim-paredit/lang/init.lua [hunk 2]
```diff
@@ -10,16 +16,28 @@ local function keys(tbl)
...
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  local lang = M.get_language_api()
+  cursor = { cursor[1] - 1, cursor[2] }
+
+  local char_under_cursor = vim.api.nvim_buf_get_text(0, cursor[1], cursor[2], cursor[1], cursor[2] + 1, {})
+  return common.included_in_table(
+    lang.whitespace_chars or default_whitespace_chars,
+    char_under_cursor[1]
+  ) or char_under_cursor[1] == ""
+end
```
Comment (julienvincent) [inline lua/nvim-paredit/lang/init.lua:41]: I think this fn should live in utils somewhere rather

File: lua/nvim-paredit/init.lua [hunk 3]
```diff
@@ -6,6 +6,8 @@ local lang = require("nvim-paredit.lang")
 
 local M = {
   api = require("nvim-paredit.api"),
+  form = require("nvim-paredit.api.form"),
```
Comment (julienvincent) [inline lua/nvim-paredit/init.lua:9]: I don't think `form` is the best API namespace for this. Maybe `wrap`?

File: lua/nvim-paredit/lang/clojure.lua (hunk 1 above)
Comment (armed) [reply lua/nvim-paredit/lang/clojure.lua:48]: `:foo/my-keyword` is represended as
- kwd_lit `:` - whole node range
  - kwd_ns `foo` - namespace token range
  - kwd_name `my-keyword` - name token tange

And depending on where is your cursor it returns on of this three nodes.
say if our cursor is on ns or name, then `get_node_root` returns `kwd_lit` and we are fine. But it won't work if our cursor is on `:`

File: lua/nvim-paredit/lang/init.lua (hunk 2 above)
Comment (armed) [reply lua/nvim-paredit/lang/init.lua:41]: agree

File: lua/nvim-paredit/lang/clojure.lua (hunk 1 above)
Comment (armed) [reply lua/nvim-paredit/lang/clojure.lua:48]: `:foo/my-keywo|rd` - node is `kwd_name`
`:fo|o/my-keyword` - node is `kwd_ns`
`|:foo/my-keywo|rd` - node is `kwd_lit`

`element_lit` always returns `kwd_lit`

Comment (armed) [reply lua/nvim-paredit/lang/clojure.lua:48]: Here `(:foo/my-ke|yword)` `element_lit` returns `kwd_lit`, but not the parent node which is parenthesis

Comment (julienvincent) [reply lua/nvim-paredit/lang/clojure.lua:48]: If the cursor is on `:` (`kwd_lit`) then `get_node_root` _should_ return `kwd_lit`. If not then it's a bug. Or perhaps I misunderstand you/the requirement

Comment (julienvincent) [reply lua/nvim-paredit/lang/clojure.lua:48]: `get_node_root` should also always return `kwd_lit` in all your above examples.

File: lua/nvim-paredit/init.lua (hunk 3 above)
Comment (armed) [reply lua/nvim-paredit/init.lua:9]: don't disagree, rename incoming :)

File: lua/nvim-paredit/lang/clojure.lua (hunk 1 above)
Comment (armed) [reply lua/nvim-paredit/lang/clojure.lua:48]: Ok, you are right, I didn't get it from the description of `get_node_root` and was too lazy to really test stuff with it :). Removing `element_lit`.

Comment (julienvincent) [reply lua/nvim-paredit/lang/clojure.lua]: Yea the fn comment isn't particularly clear on what it's for/does haha. Cool

**Review by julienvincent — APPROVED** [review]

**Pull request merged**